### PR TITLE
#002 create register screen

### PR DESCRIPTION
**Pull request by halfangel37**
## Change Description 👋

I created a register form with First Name , Last Name , Email (UserName), Password, I also do the validate with the input, the form make an API request to the specific link, the form also have responsive on different device.
Update: split Register screen to views/RegisterPage.vue and components/RegisterForm.vue, update store structure, 
note: in the video the button not disabled when signup success but i've fixed it.
Update: Im using mapGetter to get data from state, refactor syntax to optimize

## Demo 📺
 Demo:
https://www.loom.com/share/25c7d4566d2e405195b66039300a3f5f
Update new Demo for new successful message:
https://www.loom.com/share/cab3800d399941279fc01a5afca5d7d6
Register screen with layout (click signup on login page to move to signup page) :
image

### CODE REVIEW

**Review by truonghungit — CHANGES_REQUESTED** [review]
Nice work @halfangel37
I've left quite a few comments so let me know if anything is unclear.
Also, please 
- Add me as a reviewer when creating a PR
- Follow Git branch naming convention
- Follow Git commit message convention

**Comment by truonghungit** [comment]
> it worked without bug before , i dont know where the 403 status come from.

@halfangel37  Let investigate and fix it, you shouldn't leave this message in the pull request. You can ask me, or the team to support

File: src/store.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  },
+  actions: {
+    register({ commit }, credentials) {
+      return HTTP.post("/auth/register", credentials)
+        .then(({ data }) => {
+          console.log("user data is", data);
+          commit("SET_USER_DATA", data);
+        })
+        .catch((error) => {
+          console.log(error.response);
```
Comment (truonghungit) [inline src/store.js:30]: When the user registers success/failure, please check the response and present a message to them. Also remove console log here

image

File: src/store.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        "Authorization"
+      ] = `Bearer ${userData.token}`;
+    },
+  },
+  actions: {
+    register({ commit }, credentials) {
+      return HTTP.post("/auth/register", credentials)
+        .then(({ data }) => {
+          console.log("user data is", data);
+          commit("SET_USER_DATA", data);
```
Comment (truonghungit) [inline src/store.js:27]: Don't need to save user data

File: src/store.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import Vuex from "vuex";
+import axios from "axios";
+Vue.use(Vuex);
+export const HTTP = axios.create({
+  baseURL: `http://intern-2021.southeastasia.cloudapp.azure.com:4000/api/v1`,
+});
+
+export default new Vuex.Store({
+  state: {
+    user: null,
```
Comment (truonghungit) [inline src/store.js:11]: for what? angel

File: src/views/Register.vue [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+      (value) =>
+        value.indexOf(".") - value.indexOf("@") > 1 ||
+        "Email should contain a valid domain.",
+      (value) => value.includes(".") || "Email should include a period symbol.",
+      (value) =>
+        value.indexOf(".") <= value.length - 3 ||
+        "Email should contain a valid domain extension.",
+    ],
+    rules: {
+      required: (value) => !!value || "The field is required.",
```
Comment (truonghungit) [inline src/views/Register.vue:87]: image

I passed your validation by space

File: src/views/Register.vue [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+          />
+          <v-btn
+            block
+            color="#FFBF3F"
+            @click="validateForm"
+            type="submit"
+            :disabled="!formValidity"
+          >
+            SIGN UP
+          </v-btn>
```
Comment (truonghungit) [inline src/views/Register.vue:53]: While calling the API, please deactivate the **Sign up** button and display the loading indicator.

File: src/store.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import Vue from "vue";
```
Comment (truonghungit) [inline src/store.js:1]: When our app grows in scale, the `src/store.js` file might get very large. Let's break it down into modules. 
In my view, you should have a file structure as below

image

I will explain more detail about this structure in zoom call tmr.

**Comment by truonghungit** [comment]
The UI is not responsive on mobile. check again
image

File: src/register/components/RegisterForm.vue [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    validateForm() {
+      this.$refs.signUpForm.validate();
+    },
+    onSubmit() {
+      this.$emit("user-infor", this.userInfor);
+    },
+  },
+  computed: {
+    errorMessage() {
+      return this.$store.state.registerModule.errorMessage;
```
Comment (truonghungit) [inline src/register/components/RegisterForm.vue:90]: The Register form is a presentation component, it shouldn't access to state, dispatch an action or call API
You should get the error message in the `RegisterPage` component then you can
- display the error message in the `RegisterPage` component OR
- pass the error message to `RegisterForm` via props

and please use Getters to get data from state
https://vuex.vuejs.org/guide/getters.html#property-style-access

**Review by truonghungit — CHANGES_REQUESTED** [review]

File: src/register/store/Register.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
+import { HTTP } from "../../http-common";
+import router from "../../router";
+
+export const registerModule = {
```
Comment (truonghungit) [inline src/register/store/Register.js:4]:
```suggestion
export default  {
```

File: src/register/views/RegisterPage.vue [hunk 9]
```diff
@@ -0,0 +1,54 @@
+<template>
+  <v-main id="register-page">
+    <v-container id="form-container">
+      <v-row class="register-row" justify="center">
+        <v-col cols="12" sm="12" md="8" lg="8">
+          <h1>Register</h1>
+          <register-form @user-infor="register"></register-form>
```
Comment (truonghungit) [inline src/register/views/RegisterPage.vue:7]: The name of the event should be a verb. 
```suggestion
          <RegisterForm @on-submit="onSubmit" />
```

**Review by truonghungit — CHANGES_REQUESTED** [review]

File: src/register/store/Register.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    errorMessage: "null",
+  },
+  mutations: {
+    SET_ERROR_MESSAGE(state, message) {
+      state.errorMessage = message;
+    },
+  },
+  actions: {
+    register({ commit }, credentials) {
+      commit("SET_ERROR_MESSAGE", "null");
```
Comment (luantc13) [inline src/register/store/Register.js]: `"null"` is a string with `null` content, pls use `null` or `""`.

**Review by luantc13 — APPROVED** [review]
Nice work!

**Pull request merged**